Build fails for lib when importing a publishable workspace library into another publishable workspace library

Build fails when we import secondLib into firstLib and try to perform build on first lib.

Setup:
nx-poc (react app)
firstLib (publishable workspace lib wherein secondLib is imported)
secondLib (publishable wrokspace lib)

![image](https://user-images.githubusercontent.com/85344164/126630802-98cef3dc-bf65-4411-b98f-e74810fdc447.png)



## Current Behavior
We get error when we build firstLib 
nx build first-lib --with-deps --skip-nx-cache

Error/output:
Compiling TypeScript files for project "first-lib"...
libs/first-lib/src/lib/first-lib.ts:1:27 - error TS6059: File 'C:/RnD/nx-poc/libs/second-lib/src/index.ts' is not under 'rootDir' 'libs/first-lib'. 'rootDir' is expected to contain all source files.

1 import { secondLib } from "@nx-poc/secondLib";

## Expected Behavior
Build should be successful as we have imported second lib and second-lib is already built successfully.

## Steps to Reproduce

1. npx create-nx-workspace nx-poc
2. nx generate @nrwl/workspace:library --name=firstLib --buildable --importPath=@nx-poc/firstLib --no-interactive
3. nx generate @nrwl/workspace:library --name=secondLib --buildable --importPath=@nx-poc/secondLib --no-interactive
4. Import secondLib  in firstLib i.e. import { secondLib } from "@nx-poc/secondLib";
5. Build firstLib i.e nx build first-lib --with-deps --skip-nx-cache
6. Build fails with below mentioned error.

### Failure Logs
![image](https://user-images.githubusercontent.com/85344164/126629444-d6d3e765-0b25-4a64-bce5-cbb0e7a288d0.png)

### Environment
![image](https://user-images.githubusercontent.com/85344164/126629644-06204e78-eb4a-4405-bf45-1f9948b8ed24.png)


**Known workaround:**

https://github.com/nrwl/nx/issues/602
As per above mentioned issue build gets successful if we add dist under paths as below

"paths": {
      "@nx-poc/firstLib": [
        "libs/first-lib/src/index.ts"
      ],
      "@nx-poc/secondLib": [
        "dist/libs/second-lib",
        "libs/second-lib/src/index.ts"
      ]
    }

_But this looks like a workaround as it assumes that second-lib is always present in dist folder and throws error if its not present._

--- a/packages/workspace/src/executors/tsc/tsc.impl.spec.ts
+++ b/packages/workspace/src/executors/tsc/tsc.impl.spec.ts
@@ -1,31 +1,44 @@
+jest.mock('../../core/project-graph');
+jest.mock('../../utilities/assets');
+jest.mock('../../utilities/buildable-libs-utils');
+jest.mock('../../utilities/fileutils');
+jest.mock('../../utilities/typescript/compilation');
+
 import { ExecutorContext } from '@nrwl/devkit';
 import { join } from 'path';
 import { copyAssets } from '../../utilities/assets';
+import {
+  calculateProjectDependencies,
+  checkDependentProjectsHaveBeenBuilt,
+  createTmpTsConfig,
+} from '../../utilities/buildable-libs-utils';
 import { readJsonFile, writeJsonFile } from '../../utilities/fileutils';
 import { compileTypeScript } from '../../utilities/typescript/compilation';
 import { TypeScriptExecutorOptions } from './schema';
 import { tscExecutor } from './tsc.impl';
-
-const defaultPackageJson = { name: 'workspacelib', version: '0.0.1' };
-jest.mock('../../utilities/fileutils', () => ({
-  ...jest.requireActual<any>('../../utilities/fileutils'),
-  writeJsonFile: jest.fn(),
-  readJsonFile: jest.fn(() => ({ ...defaultPackageJson })),
-}));
-const readJsonFileMock = readJsonFile as jest.Mock<any>;
-jest.mock('../../utilities/typescript/compilation');
-const compileTypeScriptMock = compileTypeScript as jest.Mock<{
-  success: boolean;
-}>;
-jest.mock('../../utilities/assets');
 
 describe('executor: tsc', () => {
   const assets = ['some-file.md'];
   let context: ExecutorContext;
   let normalizedOptions: TypeScriptExecutorOptions;
   let options: TypeScriptExecutorOptions;
+  const defaultPackageJson = { name: 'workspacelib', version: '0.0.1' };
+  const compileTypeScriptMock = compileTypeScript as jest.Mock;
+  const readJsonFileMock = readJsonFile as jest.Mock;
 
   beforeEach(() => {
+    jest.clearAllMocks();
+
+    (calculateProjectDependencies as jest.Mock).mockImplementation(() => ({
+      target: { data: { root: 'libs/workspacelib' } },
+      dependencies: [],
+    }));
+    (createTmpTsConfig as jest.Mock).mockImplementation(
+      () => '/my-app/tsconfig.app.generated.json'
+    );
+    (checkDependentProjectsHaveBeenBuilt as jest.Mock).mockReturnValue(true);
+    readJsonFileMock.mockImplementation(() => ({ ...defaultPackageJson }));
+
     context = {
       cwd: '/root',
       root: '/root',
@@ -54,8 +67,19 @@ describe('executor: tsc', () => {
       outputPath: join(context.root, options.outputPath),
       tsConfig: join(context.root, options.tsConfig),
     };
+  });
 
-    jest.clearAllMocks();
+  it('should return { success: false } when dependent projects have not been built', async () => {
+    (calculateProjectDependencies as jest.Mock).mockImplementation(() => ({
+      target: { data: { root: 'libs/workspacelib' } },
+      dependencies: [{}],
+    }));
+    (checkDependentProjectsHaveBeenBuilt as jest.Mock).mockReturnValue(false);
+
+    const result = await tscExecutor(options, context);
+
+    expect(result).toEqual({ success: false });
+    expect(compileTypeScriptMock).not.toHaveBeenCalled();
   });
 
   it('should return typescript compilation result', async () => {

--- a/packages/workspace/src/executors/tsc/tsc.impl.ts
+++ b/packages/workspace/src/executors/tsc/tsc.impl.ts
@@ -1,11 +1,14 @@
-import { ExecutorContext, logger, normalizePath } from '@nrwl/devkit';
+import { ExecutorContext, normalizePath } from '@nrwl/devkit';
 import { basename, dirname, join, relative } from 'path';
+import { readCachedProjectGraph } from '../../core/project-graph';
 import { copyAssets } from '../../utilities/assets';
-import { readJsonFile, writeJsonFile } from '../../utilities/fileutils';
 import {
-  compileTypeScript,
-  TypescriptWatchChangeEvent,
-} from '../../utilities/typescript/compilation';
+  calculateProjectDependencies,
+  checkDependentProjectsHaveBeenBuilt,
+  createTmpTsConfig,
+} from '../../utilities/buildable-libs-utils';
+import { readJsonFile, writeJsonFile } from '../../utilities/fileutils';
+import { compileTypeScript } from '../../utilities/typescript/compilation';
 import { TypeScriptExecutorOptions } from './schema';
 
 export async function tscExecutor(
@@ -13,7 +16,36 @@ export async function tscExecutor(
   context: ExecutorContext
 ) {
   const normalizedOptions = normalizeOptions(options, context);
-  const projectRoot = context.workspace.projects[context.projectName].root;
+  // const projectRoot = context.workspace.projects[context.projectName].root;
+
+  const projectGraph = readCachedProjectGraph();
+  const { target, dependencies } = calculateProjectDependencies(
+    projectGraph,
+    context.root,
+    context.projectName,
+    context.targetName,
+    context.configurationName
+  );
+  const projectRoot = target.data.root;
+
+  if (dependencies.length > 0) {
+    const areDependentProjectsBuilt = checkDependentProjectsHaveBeenBuilt(
+      context.root,
+      context.projectName,
+      context.targetName,
+      dependencies
+    );
+    if (!areDependentProjectsBuilt) {
+      return { success: false };
+    }
+
+    normalizedOptions.tsConfig = createTmpTsConfig(
+      join(context.root, options.tsConfig),
+      context.root,
+      projectRoot,
+      dependencies
+    );
+  }
 
   // this has to happen first so the folder is created where the assets are copied into
   const result = compileTypeScript({

